Finished LED Subsystem

avakimc opened the pull request:
- made setSpacing method
- made setColor method that sets LED color using RGB
- made setLight method that sets the light based on passed in LightCodes enum
- made toggle methods for all light options in LightCodes

balghane committed:
Add Lights subsystem

avakimc committed:
Finished Light Subsystem

- made toggle methods for all lightCodes including OFF
- made setLight method that changes the light to the one passed in
- made setColor method that is used in setLight and toggle methods to change the RGB of the led
- made setSpacing method that sets the spacing between leds in the strip so not all are lit up at once (default is all are lit at once)

balghane commented on src/main/java/frc/robot/subsystems/Lights.java:
```diff
@@ -0,0 +1,142 @@
+
+public class Lights extends SubsystemBase {
+  private LightCodes currentLightStatus = LightCodes.OFF;
+  private AddressableLED m_led = new AddressableLED(RobotMap.LED_PWM_PORT);
+  private AddressableLEDBuffer m_ledBuffer;
```
Let's not use this `m_` convention and just call things what they are in camelCase

balghane commented on src/main/java/frc/robot/subsystems/Lights.java:
```diff
@@ -0,0 +1,142 @@
+  private LightCodes currentLightStatus = LightCodes.OFF;
+  private AddressableLED m_led = new AddressableLED(RobotMap.LED_PWM_PORT);
+  private AddressableLEDBuffer m_ledBuffer;
+  private int spacing = 1;
```
Does `1` mean every `1` light is on, or `1` space between every on light? Feel free to clarify in a comment

avakimc replied:
We added a comment as well, but a spacing value of 1 would mean every LED is on

balghane commented on src/main/java/frc/robot/subsystems/Lights.java:
```diff
@@ -0,0 +1,142 @@
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.RobotMap;
+
+public class Lights extends SubsystemBase {
```
Comment for the class overall

balghane commented on src/main/java/frc/robot/subsystems/Lights.java:
```diff
@@ -0,0 +1,142 @@
+    m_led.start();
+  }
+
+  public void toggleReadyToScoreLight() {
+    if (currentLightStatus == LightCodes.READY_TO_SCORE) {
+      currentLightStatus = LightCodes.OFF;
+    } else {
+      currentLightStatus = LightCodes.READY_TO_SCORE;
+    }
+    setColor(currentLightStatus);
+    m_led.setData(m_ledBuffer);
+    m_led.start();
+  }
```
You could simplify all of these and have one `toggleCode` function that takes an enum, right?

balghane commented on src/main/java/frc/robot/subsystems/Lights.java:
```diff
@@ -0,0 +1,142 @@
+    m_led.setLength(m_ledBuffer.getLength());
+  }
+
+  public enum LightCodes {
```
Asher named this one but this class name should be singular

balghane commented on src/main/java/frc/robot/subsystems/Lights.java:
```diff
@@ -0,0 +1,142 @@
+    int g = 0;
+    int b = 0;
+
+    if (light == LightCodes.CONE || light == LightCodes.CUBE || light == LightCodes.WORKING) {
+      r = 255;
+    }
+    if (light == LightCodes.CONE || light == LightCodes.GAME_OBJECT || light == LightCodes.NO_TAG) {
+      g = 255;
+    }
+    if (light == LightCodes.CUBE || light == LightCodes.READY_TO_SCORE) {
+      b = 255;
+    }
+    if (light == LightCodes.OFF) {
+      r = 0;
+      g = 0;
+      b = 0;
+    }
```
I think this would be much more readable as a big switch statement?

balghane replied:
And even if it's more lines of code, it'll be more readable if you redundantly define all three variables in every case of the switch.

balghane commented on src/main/java/frc/robot/subsystems/Lights.java:
```diff
@@ -0,0 +1,142 @@
+    if (light == LightCodes.CONE || light == LightCodes.CUBE || light == LightCodes.WORKING) {
+      r = 255;
+    }
+    if (light == LightCodes.CONE || light == LightCodes.GAME_OBJECT || light == LightCodes.NO_TAG) {
```
What about blinking?

avakimc committed:
Revision to Lights subsystem

-switched out the massive if block for a switch statement
-removed individual toggles, added one toggle with a parameter
-added blink for NO_TAG

balghane commented on src/main/java/frc/robot/subsystems/Lights.java:
```diff
@@ -112,31 +54,66 @@ public void setSpacing(int s) {
-  public void periodic() {}
+  public void periodic() {
+    if (currentLightStatus == LightCode.NO_TAG) {
+      if (blinkingTimer == 500) {
+        toggleCode(LightCode.NO_TAG);
+        blinkingTimer = 0;
+      }
+    }
+    blinkingTimer += 50;
```
Not sure what 50 and 500 are here. Really what you're doing here is switching every 10 cycles. So you might want a variable equal to 10 with some descriptive name (blinking period or something) and use it that way

balghane commented on src/main/java/frc/robot/subsystems/Lights.java:
```diff
@@ -112,31 +54,66 @@ public void setSpacing(int s) {
+  public void periodic() {
+    if (currentLightStatus == LightCode.NO_TAG) {
+      if (blinkingTimer == 500) {
+        toggleCode(LightCode.NO_TAG);
```
Won't this turn `currentLightStatus` to OFF, thereby only flashing once?

avakimc committed:
Final Revisions to Light Subsystem

- compared blinkingTimer to blinkingPeriod, which is equal to 10 cycles (500 ms)
- created a boolean, currentlyBlinking that is true when the light status should be blinking (ex. NO_TAG)

avakimc committed:
Updated Lights Subsystem to Avoid Bug

* made an AddressableLEDBuffer for each color option
* made tree map to associate each LightCode with its AddressableLEDBuffer
* hopefully fixed bug so blinking works as intended
* changed length of LED to a placeholder in Constants

balghane commented on src/main/java/frc/robot/subsystems/Lights.java:
```diff
@@ -3,27 +3,74 @@
+  private TreeMap<LightCode, AddressableLEDBuffer> lightOptionsMap;
   private AddressableLED led = new AddressableLED(RobotMap.LED_PWM_PORT);
-  private AddressableLEDBuffer ledBuffer;
+  private AddressableLEDBuffer ledBufferCone;
+  private AddressableLEDBuffer ledBufferCube;
+  private AddressableLEDBuffer ledBufferGameObject;
+  private AddressableLEDBuffer ledBufferNoTag;
+  private AddressableLEDBuffer ledBufferWorking;
+  private AddressableLEDBuffer ledBufferReadyToScore;
+  private AddressableLEDBuffer ledBufferOff;
```
These don't need to be member variables because they live in the map

balghane commented on src/main/java/frc/robot/subsystems/Lights.java:
```diff
@@ -3,27 +3,74 @@
+    led.setLength(ledBufferOff.getLength());
+
+    // sets the rgb values for each of the AddressableLEDBuffers
+    for (int i = 0; i < ledBufferCone.getLength(); i = i + spacing) {
+      ledBufferCone.setRGB(i, 255, 255, 0);
+    }
```
Can you add a helper function that fills a given buffer with provided RGB values so that we don't have a bazillion for loops

balghane commented on src/main/java/frc/robot/subsystems/Lights.java:
```diff
@@ -59,74 +104,19 @@ public void setSpacing(int s) {
       if (blinkingTimer >= blinkingPeriod) {
-        toggleCode(LightCode.NO_TAG);
-        blinkingTimer = 0;
+        if (blinkingColorOn) {
+          led.setData(ledBufferOff);
+        } else {
+          led.setData(ledBufferNoTag);
+        }
```
This seems flipped

avakimc committed:
Helper Function for Light Subsystem

* Helper function to set RGB values for a passed in buffer
* Removed AddressableLEDBuffer member variables

balghane commented:
/format

github-actions[bot] committed:
Formatting fixes
